### ISSUE TITLE
Git diff shows difference on CustomLabels.xml, but sgd does not add the file in the package.xml

### ISSUE THREAD

**Issue by AllanOricil** [scolladon/sfdx-git-delta#480]
Issue verification check:

- [X] is the current repository fully deployable at the commit SHA provided with the 'from' parameter of the command?

## What is the problem?

---

Git diff shows difference on `CustomLabels.xml`, but sgd does not add it in the `package.xml`

In this next image you can see that the output of `git diff origin/main..HEAD -- core/common/main/default/labels/CustomLabels.labels-meta.xml ` contains changes to `CustomLabels`
![image](https://user-images.githubusercontent.com/55927613/221186807-3067f2ce-3f02-4472-ab29-5c429ce8ac1a.png)

The following image shows that main and feature local branches are up to date.
![image](https://user-images.githubusercontent.com/55927613/221188033-9dd22192-8d14-4f28-9420-ebdb50a86af7.png)

In this image you can see that `sgd` did not include `CustomLabels.xml` changes in the `package.xml` when running `sfdx sgd:source:delta -f origin/main -t HEAD -o . -i .sgdignore`

![image](https://user-images.githubusercontent.com/55927613/221187495-efec4f9f-1c82-41db-911f-d0095d82f9c1.png)

In this image you can see that `sgd` did not include `CustomLabels.xml` changes in the `package.xml` even after removing `-i .sgdignore`, which proves we are not excluding the folder where the custom labels file is.

![image](https://user-images.githubusercontent.com/55927613/221187788-35f8da97-c498-4adc-b877-a6d623fe5323.png)

If I move `CustomLabels.labels-meta.xml` to a different folder (in this case I moved the file back to where it was), `sgd` includes all custom labels in the `package.xml`

![image](https://user-images.githubusercontent.com/55927613/221191678-af5c52a8-599a-4b11-8e49-3a0a87944719.png)

### What is the parameter and the value you used with it?

`sfdx sgd:source:delta -f origin/main -t HEAD ...`

### What is the expected result?

````xml
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <types>
        <members>SER_BillingAccountListViewController</members>
        <members>SER_BillingAccountListViewControllerTest</members>
        <members>SER_BillingAccountsCalloutService</members>
        <name>ApexClass</name>
    </types>
    <types>
        <members>PERM_PAYMENT_EXTENSION_TAKE_AWAY</members>
        <name>CustomPermission</name>
    </types>
    <types>
        <members>billingAccountListViewPaymentExtensionModal</members>
        <members>billingAccountListViewService</members>
        <name>LightningComponentBundle</name>
    </types>
    <types>
        <members>Payment_Extension_Take_Away</members>
        <name>PermissionSet</name>
    </types>
    <types>
        <members>BillingAccountListViewChangeToRafraennGreidslusedillPromptMessage</members>
        <members>BillingAccountListViewChangeToRafraennGreidslusedillSuccess</members>
        <members>BillingAccountListViewChangeToRafraennGreidslusedillFailure</members>
        <members>BillingAccountListViewTableChangeToRafraennGreidslusedillAction</members>
        <members>BillingAccountListViewPaymentExtensionSuccessMessage</members>
        <members>BillingAccountListViewTableBillingAccountClosedErrorMessage</members>
        <name>CustomLabels</name>
    </types>
    <version>57.0</version>
</Package>
````

### What is the actual result?

````xml
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <types>
        <members>SER_BillingAccountListViewController</members>
        <members>SER_BillingAccountListViewControllerTest</members>
        <members>SER_BillingAccountsCalloutService</members>
        <name>ApexClass</name>
    </types>
    <types>
        <members>PERM_PAYMENT_EXTENSION_TAKE_AWAY</members>
        <name>CustomPermission</name>
    </types>
    <types>
        <members>billingAccountListViewPaymentExtensionModal</members>
        <members>billingAccountListViewService</members>
        <name>LightningComponentBundle</name>
    </types>
    <types>
        <members>Payment_Extension_Take_Away</members>
        <name>PermissionSet</name>
    </types>
    <version>57.0</version>
</Package>
````

## Steps to reproduce

no idea.
I just know that when we were working with a single folder it was working fine. Then after I moved `CustomLabels.labels-meta.xml` to this folder `core/common/main/default/labels` sgd stopped seeing it.

## Execution context

---

![image](https://user-images.githubusercontent.com/55927613/221189677-86294171-f7a3-4b7f-bde9-7e231dda5d51.png)

**Comment by scolladon** [comment]
Hi @AllanOricil,

Thanks for raising this issue and thanks for contributing in making this project better!
Very clean description, easy to jump in!

First, I see the custom label files content is changed a bit, it looks like it has been prettified to limit the number of character on a row I would say.
Then the fullname is on multiple lines
I think it creates an issue here, the plugin read the fullname content in the same line and here it cannot find it.

Could you try put the fullname on the same line and tell us if it works ?

**Why the plugin need fullname to be on the same line?**
- the metadata api do not split xml tags when retrieving them
- when we implemented the algorithm the first time we did not want to use xml parser or json converter with complex diff algorithm fetching the file at both state and compare them because it was not needed with the way the metadata api works, it is more complex to engineer and it is also a bit slower

If you feel the plugin must be more "robust" and compatible with any prettier configuration we can convert this issue in enhancement and track this 👍

**Comment by AllanOricil** [comment]
@scolladon what if we run prettier on .xml files before reading them to garantee that there will never bee line breaks?  (HACK)

**Comment by AllanOricil** [comment]
@scolladon would this break sgd as well?
![image](https://user-images.githubusercontent.com/55927613/221197575-57a3e051-df65-4fc2-b728-2acd9d9e6a6f.png)

**Comment by AllanOricil** [comment]
I will open a PR by the end of the next week

**Comment by scolladon** [comment]
> @scolladon what if `sgd` runs prettier on .xml files before reading them in order to garantee that there will never bee line breaks? Prettier can be used here to force files to always follow a pattern that `sgd` expects. This will make the code bit slower, and the plugin will have to be shipped with prettier.
> 
> something like this would remove these line breaks `./node_modules/.bin/prettier --write '**/*.xml' --print-width=Infinity`
> 
> it is not a hack. it is actually a solution. It is the same thing as forcing a `schema` before deserializing data.

I agree it is a suitable solution, I feel it can be a good intermediate solution even it requires lot of work because of the new plugin architecture. Now we take file content from git on not from file system. So we would have to query the repo to get the file content, prettier infinity width on the data returned (breaking a few optimization we have on its way) and then apply our in house algorithm.

> xml/json parsers would be a proper solution? I don't believe so because the plugin is not interested in reading data based on a certain query. It only needs to know the data that changed.

I agree, but I feel there is not a lot of elegant solution to build a generic solution for this area, something that could be outsourced after.

Based on this discussion I definitely think it is a worthy enhancement, that requires some spiking and user story breaks down

**Comment by scolladon** [comment]
> I will open a PR by the end of the next week

Nice ! let's work on this ! Thanks for the contribution

**Comment by RupertBarrow** [comment]
> I agree, but I feel there is not a lot of elegant solution to build a generic solution for this area, something that could be outsourced after.
> 
> 
> 
> Based on this discussion I definitely think it is a worthy enhancement, that requires some spiking and user story breaks down

I reported a similar issue with the SFDX diff option of the DX extension for VS Code which was not triggering automatic prettifying before calculating the diff.
My issue was accepted, an enhancement Work Item is in progress.

**Comment by AllanOricil** [comment]
@scolladon I started working on this one but I could not move the ticket in the Project kamban

**Comment by AllanOricil** [comment]
I'm still learning the flow of the code. I just created a debugger config in vscode to help me with that

![image](https://user-images.githubusercontent.com/55927613/224195340-c15a1b0f-0dc1-4589-9648-c7b7ae26d971.png)

**Comment by AllanOricil** [comment]
@scolladon 
Now that I undersand how the code works, let me share my thoughts about my idea of using prettier. 

I'm thinking about creating two new hashes.


````shell
-------------------------> TIME
   ----* run prettier
   |    TO'           ----* run prettier
   |                    |      FROM'
   |                    |
--*-----------*------------
  TO              FROM

`````

Then instead I will run the same flow using TO' and FROM'. What do you think?

**Comment by scolladon** [comment]
Hi @AllanOricil 

I have just finished the [first draft of a PR](https://github.com/scolladon/sfdx-git-delta/pull/511) to address this enhancement request. Thanks to your energy and discussion about it!

Could you [test this PR](https://github.com/scolladon/sfdx-git-delta/blob/main/CONTRIBUTING.md#testing-the-plugin-from-a-pull-request) and let us know if it fix the change identification issue because of indentation on your side?

Thanks a lot for your contribution on this one

**Comment by AllanOricil** [comment]
@scolladon I felt extremely useless, and betrayed when you shared your PR. I won't probably try to help here anymore until these feelings go away.

**Comment by scolladon** [comment]
Oh @AllanOricil 😯! 
I am really sorry you felt that way...
I can tell you that it was definitely not my intention.
I apologize for any wrong you felt from my doing and thank you very much for your honest and open feedback.

I feel grateful for your contributions and our discussions here.
Thanks to you, we saw the limits of the algorithm we had.
And talking design with you made an idea emerge and gave me energy to give it a try (plus I had the week-end off and an old on going branch on this matter I was able to reuse, which made the work a lot easier than I initially thought)

You’re one of the few contributors that took some time to help us find ways to improve further the plugin, and we feel really lucky to have you in our community. So, this is just to let you know that you’re more than welcome to contribute again, at any level.
This enhancement wouldn’t have been possible without your initial contribution. The whole community will now benefits from it, so again, thank you!
